Fix the issue in CPU static policy when containers shouldn't be sched…

dastonzerg opened the pull request:
…uled on reserved CPUs.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind bug

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #115994

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

dastonzerg commented:
Hey @haircommander I have signed 10 mins ago but there is no "Look for an email indicating successful signup." email received as per https://github.com/kubernetes/community/blob/master/CLA.md#5-look-for-an-email-indicating-successful-signup . Should I wait for longer time?

haircommander commented:
can you try running
```
git commit --amend -s
git push -f origin zw/fix-kubelet-reserved-cpu
```

dastonzerg commented:
/easycla

dastonzerg commented:
/easycla

dastonzerg commented:
@haircommander Thx! after that I added `Signed-off-by: Weipeng Zheng <zweipeng@google.com>` line to the end of commit, and I recieved an email from `admin@lfx.linuxfoundation.org` about my signed PDF which has full name "Weipeng Zheng" and email zweipeng@google.com , then I ran `/easycla` as per https://github.com/kubernetes/community/blob/master/CLA.md#6-validate-your-cla but the CLA check is still not updated.

dastonzerg commented:
/assign @ffromani

dastonzerg commented:
/easycla

dastonzerg commented:
/easycla

dastonzerg committed:
Fix: Exclude reserved CPUs from shared pool.

Signed-off-by: Weipeng Zheng <zweipeng@google.com>

dastonzerg commented:
/easycla

dastonzerg commented:
/easycla

ffromani commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -201,7 +201,8 @@ func (p *staticPolicy) validateState(s state.State) error {
 		allCPUs := p.topology.CPUDetails.CPUs()
-		s.SetDefaultCPUSet(allCPUs)
+		nonReservedCPUs := allCPUs.Difference(p.reservedCPUs)
+		s.SetDefaultCPUSet(nonReservedCPUs)
```
thanks for the fix!
is the recovery flow (reading back state from cpu_manager_state) handled correctly? (see also lines 208-215 in policy_static.go)
Let's check (add tests) to side functions like GetAvailableCPUs GetAllocatableCPUs to check they return the expected values after this change

MarSik replied:
@ffromani Do you think we should make this configurable? Just to allow the old behavior for specific use cases. I just learned about another use case where this change is actually wanted, however I am worried the change will affect some of our users. Especially if this is ever backported.

The change here is small enough to allow for an `if`.

WDYT?

MarSik replied:
I think I changed my mind. The old behavior is probably possible via a simple runtime hook and it would just pollute kubelet.

ffromani commented:
/release note prevent to run containers on the cpus  part of the reserved-cpus set.

ffromani commented:
/release-note prevent to run containers on the cpus part of the reserved-cpus set.

ffromani commented:
/test

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

Tal-or commented:
/cc

swatisehgal commented:
/cc

SergeyKanzhelev commented:
It looks like https://testgrid.k8s.io/sig-node-presubmits#pr-kubelet-serial-gce-e2e-cpu-manager didn't have any sucessful runs for a while. @ffromani @swatisehgal  do you know anything about it?

ffromani commented:
> It looks like https://testgrid.k8s.io/sig-node-presubmits#pr-kubelet-serial-gce-e2e-cpu-manager didn't have any sucessful runs for a while. @ffromani @swatisehgal do you know anything about it?

will check ASAP

ffromani commented:
> > It looks like https://testgrid.k8s.io/sig-node-presubmits#pr-kubelet-serial-gce-e2e-cpu-manager didn't have any sucessful runs for a while. @ffromani @swatisehgal do you know anything about it?
> 
> will check ASAP

this is likely a contributing factor: https://github.com/kubernetes/test-infra/issues/29622 - but there's likely more to investigate

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

ffromani commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -201,16 +201,17 @@ func (p *staticPolicy) validateState(s state.State) error {
 		allCPUs := p.topology.CPUDetails.CPUs()
-		s.SetDefaultCPUSet(allCPUs)
+		nonReservedCPUs := allCPUs.Difference(p.reservedCPUs)
+		s.SetDefaultCPUSet(nonReservedCPUs)
```
after a quick pass, I'd extract a helper like
```
func (p *staticPolicy) allocatableCPUs() cpuset.CPUSet {
	return p.topology.CPUDetails.CPUs().Difference(p.reservedCPUs)
}
```
then in the calling site above I'd use
```
	// state is empty initialize
       s.SetDefaultCPUSet(p.allocatableCPUs())
       return nil
```

and then below
```
// GetAllocatableCPUs returns the total set of CPUs available for allocation.
func (p *staticPolicy) GetAllocatableCPUs(s state.State) cpuset.CPUSet {
	return p.allocatableCPUs()
}
```

Still need to check `GetAvailableCPUs()`

dastonzerg commented:
/retest

dastonzerg commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

ffromani commented:
xref: https://github.com/kubernetes/test-infra/issues/29694

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

the infra issue should be solved now

ffromani commented:
could you please rebase on top of current master branch? the `kubelet-serial-cpu-manager` issues should be fixed now, so we can run the lane against this PR

ffromani commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -239,7 +238,9 @@ func (p *staticPolicy) validateState(s state.State) error {
-	totalKnownCPUs = totalKnownCPUs.Union(tmpCPUSets...)
+	// Total CPUs in topology = shared default CPUs + assigned exclusive CPUs +
+	// reserved CPUs for system processes.
+	totalKnownCPUs = totalKnownCPUs.Union(append(tmpCPUSets, p.reservedCPUs)...)
```
I think `append` is not needed, this should work:
```
totalKnownCPUs = totalKnownCPUs.Union(tmpCPUSets, p.reservedCPUs)
```

dims commented:
/test pull-kubernetes-node-e2e-containerd-serial-ec2

dims commented:
/test pull-kubernetes-node-e2e-containerd-serial-ec2-canary

dastonzerg commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

ffromani commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -200,17 +200,16 @@ func (p *staticPolicy) validateState(s state.State) error {
 		// state is empty initialize
-		allCPUs := p.topology.CPUDetails.CPUs()
-		s.SetDefaultCPUSet(allCPUs)
+		s.SetDefaultCPUSet(p.allocatableCPUs())
```
I'm still somehow bothered by the fact we initialize the state in a function called `validateState` but we did this before so I guess we can keep it this way and refactor later

ffromani reviewed (commented):
besides the inline (nice to have, but non-blocking) comments, unit test coverage looks good. Let's run the cpumanager e2e tests to get confidence about the change.

Even though we agreed this is a bug, still we are changing a long established kubelet behavior, so extra warrant is needed

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

iholder101 commented:
/cc

ffromani commented:
I apologize I didn't have enough bandwidth (much to my regret) to properly investigate the implication of this change.
Even though there are reasons to change the code this way, I'm unsold this is the best approach. The change happens very deep in the flow and kinda implicitely, and I'd love to make it more explicit and evident, even though this will require some partial refactoring/redesign.

It's unfair to bill that redesign on this PR, so this tradeoff should be carefully evaluated; again unfortunately I haven't had time for this research.

For the same reason, this change happening quite deep in the flow and not explicitely, I'm not yet convinced the unit tests, albeit looking good, give enough confidence for the change. I'd like to have runs of the cpumanager and topoogy manager (and perhaps more, like podresources) e2e lanes to confirm all the major flows works as intended.

The rationale for this caution is that, albeit this is acknowledged as a bug, we are changing a longstanding behavior of the kubelet, and we need a good estimation of the extent of the consequences.

iancoolidge commented:
> I apologize I didn't have enough bandwidth (much to my regret) to properly investigate the implication of this change. Even though there are reasons to change the code this way, I'm unsold this is the best approach. The change happens very deep in the flow and kinda implicitely, and I'd love to make it more explicit and evident, even though this will require some partial refactoring/redesign.
> 
> It's unfair to bill that redesign on this PR, so this tradeoff should be carefully evaluated; again unfortunately I haven't had time for this research.
> 
> For the same reason, this change happening quite deep in the flow and not explicitely, I'm not yet convinced the unit tests, albeit looking good, give enough confidence for the change. I'd like to have runs of the cpumanager and topoogy manager (and perhaps more, like podresources) e2e lanes to confirm all the major flows works as intended.
> 
> The rationale for this caution is that, albeit this is acknowledged as a bug, we are changing a longstanding behavior of the kubelet, and we need a good estimation of the extent of the consequences.

So, what does this mean? you are blocking the CL because you don't have time to think of a better way to approach it, despite some vague concerns about it? is there anyone that can spend some time on this then?

ffromani commented:
> > I apologize I didn't have enough bandwidth (much to my regret) to properly investigate the implication of this change. Even though there are reasons to change the code this way, I'm unsold this is the best approach. The change happens very deep in the flow and kinda implicitely, and I'd love to make it more explicit and evident, even though this will require some partial refactoring/redesign.
> > It's unfair to bill that redesign on this PR, so this tradeoff should be carefully evaluated; again unfortunately I haven't had time for this research.
> > For the same reason, this change happening quite deep in the flow and not explicitely, I'm not yet convinced the unit tests, albeit looking good, give enough confidence for the change. I'd like to have runs of the cpumanager and topoogy manager (and perhaps more, like podresources) e2e lanes to confirm all the major flows works as intended.
> > The rationale for this caution is that, albeit this is acknowledged as a bug, we are changing a longstanding behavior of the kubelet, and we need a good estimation of the extent of the consequences.
> 
> So, what does this mean? you are blocking the PR because you don't have time to think of a better way to approach it, despite some vague concerns about it? is there anyone that can spend some time on this then?

Nope, I'm not blocking: I don't want and I can't. I'm just not LGTM'ing yet from my side (I'm just a reviewer here), and I explained why I'm not fully convinced about the approach, where my concerns stem from and what I think it would help gain confidence in the change.  But that's me, just one reviewer. Others can very much add their LGTM and approve if they are happy.

SergeyKanzhelev commented:
@klueska what is your position here?

swatisehgal reviewed (commented):
@dastonzerg Thanks for your work on this.

I left a few comments. This is a partial review, will continue reviewing the unit tests tomorrow.

Also, we would need to make updates to e2e tests because we have tests that check cpus assigned to non-guaranteed pods e.g. https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/cpu_manager_test.go#L252 and https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/cpu_manager_test.go#L289. Since we are removing the reserved CPUs from default cpuset, I would expect these tests to fail if we don't update the expected cpusets as the assigned CPUs would no longer match the expected cpus.

https://github.com/kubernetes/test-infra/pull/30545 has now merged so hopefully we'll have a functioning CPU Manager job very soon and my theory can be validated.

Additionally, an e2e test to determine the default cpuset for a case where we don't  run pods would be a great addition and a good way to sanity check this PR. This will also allow us to ensure that we don't cause regressions in the future.

swatisehgal commented on pkg/kubelet/cm/cpumanager/policy_static_test.go:
```diff
@@ -925,7 +928,7 @@ func TestStaticPolicyStartWithResvList(t *testing.T) {
 			stAssignments:   state.ContainerCPUAssignments{},
 			stDefaultCPUSet: cpuset.New(2, 3, 4, 5),
-			expErr:          fmt.Errorf("not all reserved cpus: \"0-1\" are present in defaultCpuSet: \"2-5\""),
+			expErr:          errors.New("current set of available CPUs \"0-11\" doesn't match with CPUs in state \"0-5\""),
```
This doesn't seem right. The reserved cpus are 0,1 so available/default should not include them and the error message. I think the error message should be something like below:
```suggestion
			expErr:          fmt.Errorf("current set of available CPUs \"2-11\" doesn't match with CPUs in state \"2-5\""),
```

Also, as per my one of previous suggestion, I think we should use continue using `fmt.Errorf` instead of `errors.New`.

swatisehgal commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -200,17 +200,16 @@ func (p *staticPolicy) validateState(s state.State) error {
-	if !p.reservedCPUs.Intersection(tmpDefaultCPUset).Equals(p.reservedCPUs) {
-		return fmt.Errorf("not all reserved cpus: \"%s\" are present in defaultCpuSet: \"%s\"",
+	if !p.reservedCPUs.Intersection(tmpDefaultCPUset).IsEmpty() {
+		return fmt.Errorf("none of reserved cpus: \"%s\" should present in defaultCpuSet: \"%s\"",
```
NIT:
```suggestion
		return fmt.Errorf("none of the reserved cpus: \"%s\" should be present in defaultCpuSet: \"%s\"",
```

swatisehgal commented on pkg/kubelet/cm/cpumanager/policy_static_test.go:
```diff
@@ -97,15 +98,15 @@ func TestStaticPolicyStart(t *testing.T) {
 			stAssignments:   state.ContainerCPUAssignments{},
 			stDefaultCPUSet: cpuset.New(0, 1),
-			expErr:          fmt.Errorf("not all reserved cpus: \"0,6\" are present in defaultCpuSet: \"0-1\""),
+			expErr:          errors.New("none of reserved cpus: \"0,6\" should present in defaultCpuSet: \"0-1\""),
```
NIT (to match with the suggestion to update the error message):
```suggestion
			expErr:          fmt.Errorf("none of the reserved cpus: \"0,6\" should be present in defaultCpuSet: \"0-1\""),
```

swatisehgal commented on pkg/kubelet/cm/cpumanager/policy_static_test.go:
```diff
@@ -171,7 +172,9 @@ func TestStaticPolicyStart(t *testing.T) {
 			}
-
+			if !policy.GetAvailableCPUs(st).Equals(st.GetDefaultCPUSet()) {
+				t.Errorf("None of reserved CPUs should present in defaultCPUSet, and therefore available CPUs should be the same as defaultCPUSet. Available CPUs: %q, defaultCPUSet: %q", policy.GetAvailableCPUs(st), st.GetDefaultCPUSet())
```
NIT:
```suggestion
				t.Errorf("None of the reserved CPUs should be present in defaultCPUSet, and therefore available CPUs should be the same as defaultCPUSet. Available CPUs: %q, defaultCPUSet: %q", policy.GetAvailableCPUs(st), st.GetDefaultCPUSet())
```

swatisehgal commented on pkg/kubelet/cm/cpumanager/policy_static_test.go:
```diff
@@ -965,7 +968,9 @@ func TestStaticPolicyStartWithResvList(t *testing.T) {
 			}
-
+			if !policy.GetAvailableCPUs(st).Equals(st.GetDefaultCPUSet()) {
+				t.Errorf("None of reserved CPUs should present in defaultCPUSet, and therefore available CPUs should be the same as defaultCPUSet. Available CPUs: %q, defaultCPUSet: %q", policy.GetAvailableCPUs(st), st.GetDefaultCPUSet())
```
Same as above:
```suggestion
				t.Errorf("None of the reserved CPUs should be present in defaultCPUSet, and therefore available CPUs should be the same as defaultCPUSet. Available CPUs: %q, defaultCPUSet: %q", policy.GetAvailableCPUs(st), st.GetDefaultCPUSet())
```

swatisehgal reviewed (commented):
Continuing review from yesterday.

swatisehgal commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-cpu-manager-kubetest2

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

now the lane should be fixed (working as expected with other test PRs)

dastonzerg commented:
@swatisehgal Hi Swati thanks for the review! I had some time today and addressed all the comments. Will run e2e tests and fix those tests accordingly as well.

dastonzerg commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-cpu-manager-kubetest2

dastonzerg commented:
@swatisehgal I see all the tests are failing with 0 CPUs excluded from container log, for e.g. `failed to match regexp "^0-3\n$" in output "1-3\n"`. The output "1-3" is from container log.

I try to find in test-infra to see that does this `kubetest2 noop` do and what is this "1-3" output about but couldn't find it. Is it printing the assigned CPU set to the pod? Want to make sure I understand what the output means so I can confirm if the test cases are failing or passing.

My commit doesn't change the behavior when reservedCPU is empty, which is the test case setup in https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/cpu_manager_test.go#L543 . But for "running a non-Gu pod", it is creating a non guaranteed pod with exactly 1 CPU in https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/cpu_manager_test.go#L258-L269, I would expect it to be assigned the default CPU set (because of https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/policy_static.go#L290) which should be `p.topology.CPUDetails.CPUs().Difference(p.reservedCPUs)` and should include all topology CPUs right, why `0` is missing... No clue about that. Need some help to understand it thx!

swatisehgal commented:
> @swatisehgal I see all the e2e test failing ones are about 0 CPUs excluded from container log, for e.g. `failed to match regexp "^0-3\n$" in output "1-3\n"`. The output "1-3" is from container log.
> 
> I try to find in test-infra to see that does this `kubetest2 noop` do and what is this "1-3" output about but couldn't find it. Is it printing the assigned CPU set to the pod? Want to make sure I understand what the output means so I can confirm if the test cases are failing or passing.
> 
> My commit doesn't change the behavior when reservedCPU is empty, which is the test case setup in https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/cpu_manager_test.go#L543 . But for "running a non-Gu pod", it is creating a non guaranteed pod with exactly 1 CPU in https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/cpu_manager_test.go#L258-L269, I would expect it to be assigned the default CPU set (because of https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/policy_static.go#L290) which should be `p.topology.CPUDetails.CPUs().Difference(p.reservedCPUs)` and should include all topology CPUs right, why `0` is missing... No clue about that. Need some help to understand it thx!

When CPU Manager is enabled, we have to ensure that either `kube-reserved` or `system-reserved` are specified. This is to ensure that the default pool never becomes empty. [ReservedSystemCPUs option](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#explicitly-reserved-cpu-list) is to explicitly set reserved CPUs and takes precedence over CPUs defined by `--kube-reserved` and `--system-reserved` options. 
Even thought [ReservedSystemCPUs  is empty](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/cpu_manager_test.go#L543) (as you pointed our earlier), please note that `kube-reserved` is set being [here](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/cpu_manager_test.go#L203-L214) in the tests.
The key to understanding the failure is in the fact that reserved CPUs is taken with its integral value in ascending order by physical core ID. You can find the code [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/cpu_manager.go#L179-L191) and this has also been documented [here](https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/#static-policy).

So, essentially CPUID "0" is reserved and your PR removes it from the default pool which explains the failing e2e tests.

Now, in doing all the digging to get to the bottom of this, it became very clear to me that keeping the reserved CPUs as part of the default pool was a very intentional design decision. It was done to avoid making the default pool completely empty and has been mentioned documented both in the code AND in the documentation. It gave us wiggle room in the default pool for best-effort, burstable and gu pods with non-integral request. As a consequence of this change, in case all the CPUs are exclusively allocated to Gu-pods requesting integral CPUs ( which is something we can expect to happen as we want maximize CPU utilization), pods that are not explicitly requesting CPUs could end up being evicted OR we would have to violate guarantee of exclusivity to prevent them from being evicted!! With reserved CPUs being part of the default pool, we had a buffer and a built-in mechanism to avoid this from happening and if we merge this PR we won't have that anymore.

Also adding the content directly from documentation:

"Note: The kubelet requires a CPU reservation greater than zero be made using either --kube-reserved and/or --system-reserved or --reserved-cpus when the static policy is enabled. **This is because zero CPU reservation would allow the shared pool to become empty.**"

With this PR, we are saying that we are okay if the default pool is exhausted. I am realizing that the implications of this PR are way more far-reaching than I thought so let's make sure we have communicated this explicitly and take our next steps based on that.

iancoolidge commented:
Hi @swatisehgal , we have gone through the process of validating issue #115994 in SIG-NODE meetings already, and it was accepted as a bug. If you would like to relitigate that, please have at it, but let's do throug that issue to avoid making this PR comment history confusing

swatisehgal commented:
@iancoolidge Sure, I was one of the attendees in the SIG Node meeting and expressed my opinion that that this issue is indeed a bug. The concerns I have raised here are not pertaining to the issue but the fix and the impact that would have.

I will defer the final LGTM and approval to @SergeyKanzhelev @klueska.

dastonzerg commented:
@swatisehgal Got it, so it is like the cpu mgr test doesn't allow empty reserved CPU at the moment, so it reserves "200m" in https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/cpu_manager_test.go#L203-L214 and then bumped up to ceiling int(1) in https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/cpu_manager.go#L191 .

I think we should reclarify what the commit is trying to fix here and what is the difference in system behavior:
Without the change:
* Reserved CPUs are part of the shared CPU pool, and shared CPU pool is for non gu pods, exclusive CPU pool is for gu pods.
* We want to make sure shared CPU pool is never empty, otherwise there will be no CPUs for non-gu pods to run. So we add the note `Note: The kubelet requires a CPU reservation greater than zero be made using either --kube-reserved and/or --system-reserved or --reserved-cpus when the static policy is enabled. This is because zero CPU reservation would allow the shared pool to become empty.` also the panic code https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/cpu_manager.go#L179-L186 .

Now, if non gu pods can run on those reserved CPUs, then it defeats the purpose for "reserving CPUs for system processes" right (see https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#explicitly-reserved-cpu-list), as we cannot tune those CPUs and expect them are actually "reserved for system processes" and not for k8s pods. So we want to fix this issue by not making non gu pods to run in reserved CPUs, and we can do so by removing reserved CPUs from shared pool, so neither gu nor non-gu pods can use those CPUs.

Regarding the "shared pool cannot be empty" problem, I think after the commit reserved CPUs can be empty, so we can just remove https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/cpu_manager.go#L179-L186 . And later on if a gu pod wants to take some exclusive CPUs away from shared pool in https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/policy_static.go#L348, it will just fail by CPU accumulator's check if we change the `>` to `>=` https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/cpu_assignment.go#L420-L422. So shared pool can never be empty we add `=` to consider the shared pool empty case.

In summary there are 2 problems we are trying to deal with:
* non-gu pods shouldn't run on reserved CPUs
* Shared CPU pool should never be empty

dastonzerg commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-cpu-manager-kubetest2

dastonzerg commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-cpu-manager-kubetest2

dastonzerg commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-cpu-manager-kubetest2

dastonzerg commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-cpu-manager-kubetest2

dastonzerg commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-cpu-manager-kubetest2

dastonzerg commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-cpu-manager-kubetest2

dastonzerg commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-cpu-manager-kubetest2

jonyhy96 commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -255,7 +256,7 @@ func (p *staticPolicy) GetAllocatableCPUs(s state.State) cpuset.CPUSet {
 // GetAvailableCPUs returns the set of unassigned CPUs minus the reserved set.
 func (p *staticPolicy) GetAvailableCPUs(s state.State) cpuset.CPUSet {
-	return s.GetDefaultCPUSet().Difference(p.reservedCPUs)
+	return s.GetDefaultCPUSet()
```
The logic is mismatch with the comment

SergeyKanzhelev commented:
/approve cancel

as discussed at yesterday sig node meeting

swatisehgal commented:
> Regarding the "shared pool cannot be empty" problem, I think after the commit reserved CPUs can be empty, so we can just remove https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/cpu_manager.go#L179-L186 . And later on if a gu pod wants to take some exclusive CPUs away from shared pool in https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/policy_static.go#L348, it will just fail by CPU accumulator's check if we change the `>` to `>=` https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/cpu_assignment.go#L420-L422. So shared pool can never be empty if we add `=` to consider the shared pool empty case.

I reviewed the code and can see that isFailed check happens in the allocation stage and all the pods that are not guaranteed or guaranteed with non-integral CPU request are [ruled out ](https://github.com/kubernetes/kubernetes/blob/release-1.28/pkg/kubelet/cm/cpumanager/policy_static.go#L290-L294) before we get to this check and we would get to this only if we are trying to allocate exclusive CPUs only.  So I stand corrected on my comment about best effort pods (with no CPU request) in this check. I can see how this would ensure that the default pool is never empty but IIUC we get this by preventing all the CPUs from being used on the node? We essentially keep a buffer of 1 CPU always on a node. Correct?

If I understand the intention above correctly, I think we need to consider the node allocatable here too. Consider this scenario: we have a node with 8 cpus, 2 cpus are reserved on the node =>  6 allocatable CPUs as per https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/policy_static.go#L252. The nodeAllocatable amount (6) is advertised on the node and subsequently used by the scheduler to perform scheduling decisions. Now, if we have a guaranteed pod requesting 6 CPUs, the scheduler would consider the aforementioned node a suitable candidate for the placement of the pod but if it is scheduled on the node it would fail at admission time. 

> In summary there are 2 problems we are trying to deal with:
> 
> * non-gu pods shouldn't run on reserved CPUs
> * Shared CPU pool should never be empty

I want to share a very compelling point made by @MarSik in our discussion yesterday about default pool not being empty:

```
An empty value indicates that the cgroup is using the same setting as the nearest cgroup ancestor with a non-empty "cpuset.cpus" or all the available CPUs if none is found.
```
This is copied verbatim from [cgroup-v2 documentation](https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#cpuset-interface-files) and essentially means that if we try to make the defaultset empty it would inherit cgroup from its nearest non-empty ancestor or get all the CPUs which in the worst case can jeopardize exclusivity guarantees of all the pinned pods. I couldn't figure out if the implications are same in case of cgroupsv1 or not. But with [cgroupsv2](https://kubernetes.io/docs/concepts/architecture/cgroups/#cgroup-v2) supported as a stable feature in Kubernetes we can't ignore the fact that even if we try to make the default pool empty it could completely backfire. I wonder if this was the very reason for ensuring that the default pool is never empty from the get go.

jonyhy96 commented:
> /approve cancel
> 
> as discussed at yesterday sig node meeting

Can you provide the discussion conclusion? Why was this proposal canceled

iancoolidge commented:
I hope we can find some time to discuss this at sig-node today, I will only be able to be in for a few minutes due to some overlapping meetings.

I appreciate the reference to cgroups v2 documentation.

To me, if someone exhausts the shared pool with exclusive allocations, that is WAI, and should be a supported thing. I don't think that anyone is arguing against that point, but I am simply stating my assumption.

Now, as exclusive allocations consume CPUs out of the shared pool, and the shared pool becomes empty, I hear the argument about cgroups v2 behavior: documentation seems to indicate that the parent cgroups resources will be used instead.

If that's the case, couldn't we simply just have a specifc check to see if the `cpuset.cpus` is empty, and fail the allocation that way? It seems like a relatively straightforward problem to solve.

dastonzerg commented:
> > Regarding the "shared pool cannot be empty" problem, I think after the commit reserved CPUs can be empty, so we can just remove https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/cpu_manager.go#L179-L186 . And later on if a gu pod wants to take some exclusive CPUs away from shared pool in https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/policy_static.go#L348, it will just fail by CPU accumulator's check if we change the `>` to `>=` https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/cpu_assignment.go#L420-L422. So shared pool can never be empty if we add `=` to consider the shared pool empty case.
> 
> I reviewed the code and can see that isFailed check happens in the allocation stage and all the pods that are not guaranteed or guaranteed with non-integral CPU request are [ruled out ](https://github.com/kubernetes/kubernetes/blob/release-1.28/pkg/kubelet/cm/cpumanager/policy_static.go#L290-L294) before we get to this check and we would get to this only if we are trying to allocate exclusive CPUs only. So I stand corrected on my comment about best effort pods (with no CPU request) in this check. I can see how this would ensure that the default pool is never empty but IIUC we get this by preventing all the CPUs from being used on the node? We essentially keep a buffer of 1 CPU always on a node. Correct?
> 
> If I understand the intention above correctly, I think we need to consider the node allocatable here too. Consider this scenario: we have a node with 8 cpus, 2 cpus are reserved on the node => 6 allocatable CPUs as per https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/policy_static.go#L252. The nodeAllocatable amount (6) is advertised on the node and subsequently used by the scheduler to perform scheduling decisions. Now, if we have a guaranteed pod requesting 6 CPUs, the scheduler would consider the aforementioned node a suitable candidate for the placement of the pod but if it is scheduled on the node it would fail at admission time.
> 
> > In summary there are 2 problems we are trying to deal with:
> > 
> > * non-gu pods shouldn't run on reserved CPUs
> > * Shared CPU pool should never be empty
> 
> I want to share a very compelling point made by @MarSik in our discussion yesterday about default pool not being empty:
> 
> ```
> An empty value indicates that the cgroup is using the same setting as the nearest cgroup ancestor with a non-empty "cpuset.cpus" or all the available CPUs if none is found.
> ```
> 
> This is copied verbatim from [cgroup-v2 documentation](https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#cpuset-interface-files) and essentially means that if we try to make the defaultset empty it would inherit cgroup from its nearest non-empty ancestor or get all the CPUs which in the worst case can jeopardize exclusivity guarantees of all the pinned pods. I couldn't figure out if the implications are same in case of cgroupsv1 or not. But with [cgroupsv2](https://kubernetes.io/docs/concepts/architecture/cgroups/#cgroup-v2) supported as a stable feature in Kubernetes we can't ignore the fact that even if we try to make the default pool empty it could completely backfire. I wonder if this was the very reason for ensuring that the default pool is never empty from the get go.
@swatisehgal Thanks for the follow up discussions and context sharing! I am clear and agree on shared CPU pool should never go empty for the following reasons based our discussions so far:
* It already behaves that way, and we don't want to change the behavior to impact users who already assume it. It is somewhat of a "regression".
* We are supporting cgroupv2 and allowing shared CPU pool could jeopardize exclusivity guarantees of all the pinned pods.

I think I can address it together in this PR too. I was thinking of throwing error in admission time like in `change the > to >= https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/cpu_assignment.go#L420-L422. So shared pool can never be empty if we add = to consider the shared pool empty case.` Let's discuss if this approach is good enough or there is any other better way in today's signode meeting.

swatisehgal commented:
> I think I can address it together in this PR too. I was thinking of throwing error in admission time like in `change the > to >= https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/cpu_assignment.go#L420-L422. So shared pool can never be empty if we add = to consider the shared pool empty case.` Let's discuss if this approach is good enough or there is any other better way in today's signode meeting.

Hi, sorry I was out sick yesterday so couldn't join the discussion in SIG Node meeting and the recording hasn't been uploaded yet. Did we manage to discuss the implications of this approach from scheduling and end to end user perspective? I am referring to my point below:

>I think we need to consider the node allocatable here too. Consider this scenario: we have a node with 8 cpus, 2 cpus are reserved on the node => 6 allocatable CPUs as per https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/policy_static.go#L252. The nodeAllocatable amount (6) is advertised on the node and subsequently used by the scheduler to perform scheduling decisions. Now, if we have a guaranteed pod requesting 6 CPUs, the scheduler would consider the aforementioned node as a suitable candidate for the placement of the pod but if it is scheduled on the node it would fail at admission time. 

Was the disconnect between scheduler and kubelet brought up in the meeting? With the approach you are suggesting we are could end up in runaway pod creation scenarios similar to https://github.com/kubernetes/kubernetes/issues/84869.

jonyhy96 commented:
I think this pr already fix the issue `non-gu pods shouldn't run on reserved CPUs` which is a known bug. Better push this pr forward and open another issue to talk about `how to solve shared CPU pool should never be empty`

pacoxu commented:
Recording: https://www.youtube.com/watch?v=HdIURTQSm7Q 
for Oct 3rd, 2023 sig node weekly meeting.

bart0sh commented:
/release-note-edig ```Prevent to run containers on the cpus part of the reserved-cpus set.```

bart0sh commented:
/release-note-edit ```Prevent to run containers on the cpus part of the reserved-cpus set.```

bart0sh commented:
/priority important-longterm

bart0sh commented:
/release-note-edit ```release-note Prevent to run containers on the cpus part of the reserved-cpus set.```

ffromani commented:
/remove-lifecycle stale

SergeyKanzhelev commented:
/remove-lifecycle stale

@ffromani removing stale assuming you still want to follow up here

swatisehgal commented:
This PR is superseded by feature currently in alpha, the progress is tracked in the enhancement issue: https://github.com/kubernetes/enhancements/issues/4540 so closing this PR.

/close
